### PR TITLE
Make Team extend ForwardingAudience

### PR DESCRIPTION
**Pull request by powercasgamer**

### CODE REVIEW

**Comment by Leguan16** [comment]
you were a second faster....

**Comment by olijeffers0n** [comment]
I suppose this makes sense, general paper start and end comment things though please :)

**Comment by Leguan16** [comment]
Propper comments are missing though :)

File: patches/server/0010-Adventure.patch [hunk 1]
```diff
@@ -4895,6 +4895,21 @@ index 5540a766bc7edf8985058e29cc419c2c6b1d4428..baff11728120fda1740d5e15163760f5
...
+@@ -303,4 +360,14 @@ final class CraftTeam extends CraftScoreboardComponent implements Team {
+         return !(this.team != other.team && (this.team == null || !this.team.equals(other.team)));
+     }
+ 
++    // Paper - make Team extend ForwardingAudience
++    @Override
++    public @org.jetbrains.annotations.NotNull Iterable<? extends net.kyori.adventure.audience.Audience> audiences() {
++        java.util.List<net.kyori.adventure.audience.Audience> audiences = new java.util.ArrayList<>();
++        for (String playerName : this.team.getPlayers()) {
++            audiences.add(Bukkit.getPlayerExact(playerName));
```
Comment (lynxplay) [inline patches/server/0010-Adventure.patch]: It is very not smart to add null into that iterable.
ForwardingAudience#sendMessage for-loops over that and would promptly NPE.

File: patches/server/0010-Adventure.patch [hunk 2]
```diff
@@ -4895,6 +4895,21 @@ index 5540a766bc7edf8985058e29cc419c2c6b1d4428..baff11728120fda1740d5e15163760f5
...
      @Override
      public String getDisplayName() {
+@@ -303,4 +360,14 @@ final class CraftTeam extends CraftScoreboardComponent implements Team {
+         return !(this.team != other.team && (this.team == null || !this.team.equals(other.team)));
+     }
+ 
++    // Paper - make Team extend ForwardingAudience
++    @Override
++    public @org.jetbrains.annotations.NotNull Iterable<? extends net.kyori.adventure.audience.Audience> audiences() {
++        java.util.List<net.kyori.adventure.audience.Audience> audiences = new java.util.ArrayList<>();
```
Comment (Leguan16) [inline patches/server/0010-Adventure.patch:4906]: also `this.checkState();` is missing at the start of the method.

File: patches/server/0010-Adventure.patch (hunk 1 above)
Comment (powercasgamer) [reply patches/server/0010-Adventure.patch]: Should I do `getOfflinePlayer` instead? That would match `CraftTeam#getPlayers()`

Comment (Leguan16) [reply patches/server/0010-Adventure.patch]: Does it extend Audience?

Comment (powercasgamer) [reply patches/server/0010-Adventure.patch]: > Does it extend Audience?

nope lol i didn't say anything :p

Comment (Leguan16) [reply patches/server/0010-Adventure.patch]: I cannot reproduce this though. 
I just used a modified version of the getPlayers defined in CraftTeam:
```java
@Override
public @NotNull Iterable<? extends Audience> audiences() {
    this.checkState();

    ImmutableSet.Builder<org.bukkit.entity.Player> players = ImmutableSet.builder();
    for (String playerName : this.team.getPlayers()) {
        org.bukkit.entity.Player player = Bukkit.getPlayerExact(playerName);
        if (player != null) {
            players.add(player);
        }
    }

    return players.build();
}
```

**Review by olijeffers0n — APPROVED** [review]

**Comment by Leguan16** [comment]
can you add me as co-author? xd

File: patches/server/0010-Adventure.patch [hunk 3]
```diff
@@ -4895,6 +4895,26 @@ index 5540a766bc7edf8985058e29cc419c2c6b1d4428..baff11728120fda1740d5e15163760f5
  
      @Override
      public String getDisplayName() {
+@@ -303,4 +360,19 @@ final class CraftTeam extends CraftScoreboardComponent implements Team {
+         return !(this.team != other.team && (this.team == null || !this.team.equals(other.team)));
+     }
+ 
++    // Paper start - make Team extend ForwardingAudience
++    @Override
```
Comment (Leguan16) [inline patches/server/0010-Adventure.patch:4903]: I am not sure but we might want to state, that it doesn't necessarily contain all members of the team as audience.

**Review by Warriorrrr — APPROVED** [review]

**Review by lynxplay — APPROVED** [review]

**Review by Machine-Maker — APPROVED** [review]

**Pull request merged**